🐞 Bug libstdc++.so.6 (in Glibc) broken in x86_64 hatch m125 container

```
-- The CXX compiler identification is Clang 18.1.8
-- Detecting CXX compiler ABI info
-- The C compiler identification is Clang 18.1.8
-- The CXX compiler identification is Clang 18.1.8
-- Detecting C compiler ABI info
-- Detecting CXX compiler ABI info - failed
-- Check for working CXX compiler: /usr/local/bin/clang++
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/local/bin/clang - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Check for working CXX compiler: /usr/local/bin/clang++ - broken
CMake Error at /usr/local/share/cmake-3.29/Modules/CMakeTestCXXCompiler.cmake:73 (message):
  The C++ compiler

    "/usr/local/bin/clang++"

  is not able to compile a simple test program.

  It fails with the following output:

    Change Dir: '/usr/local/tmp/crew/ponyc.20240622074005.dir/build/build_libs/gbenchmark-prefix/src/gbenchmark-build/CMakeFiles/CMakeScratch/TryCompile-3yBLsN'
    
    Run Build Command(s): /usr/local/bin/cmake -E env VERBOSE=1 /usr/local/bin/make -f Makefile cmTC_5212d/fast
    make[4]: Entering directory '/usr/local/tmp/crew/ponyc.20240622074005.dir/build/build_libs/gbenchmark-prefix/src/gbenchmark-build/CMakeFiles/CMakeScratch/TryCompile-3yBLsN'
    /usr/local/bin/make  -f CMakeFiles/cmTC_5212d.dir/build.make CMakeFiles/cmTC_5212d.dir/build
    make[5]: Entering directory '/usr/local/tmp/crew/ponyc.20240622074005.dir/build/build_libs/gbenchmark-prefix/src/gbenchmark-build/CMakeFiles/CMakeScratch/TryCompile-3yBLsN'
    Building CXX object CMakeFiles/cmTC_5212d.dir/testCXXCompiler.cxx.o
    /usr/local/bin/clang++   -fpic  -MD -MT CMakeFiles/cmTC_5212d.dir/testCXXCompiler.cxx.o -MF CMakeFiles/cmTC_5212d.dir/testCXXCompiler.cxx.o.d -o CMakeFiles/cmTC_5212d.dir/testCXXCompiler.cxx.o -c /usr/local/tmp/crew/ponyc.20240622074005.dir/build/build_libs/gbenchmark-prefix/src/gbenchmark-build/CMakeFiles/CMakeScratch/TryCompile-3yBLsN/testCXXCompiler.cxx
    Linking CXX executable cmTC_5212d
    /usr/local/bin/cmake -E cmake_link_script CMakeFiles/cmTC_5212d.dir/link.txt --verbose=1
    /usr/local/bin/clang++ -fpic  -flto=auto  CMakeFiles/cmTC_5212d.dir/testCXXCompiler.cxx.o -o cmTC_5212d
    /usr/local/bin/ld: /usr/local/bin/../lib64/LLVMgold.so: error loading plugin: /usr/local/lib64/libstdc++.so.6: undefined symbol: strtof128
    clang++: error: linker command failed with exit code 1 (use -v to see invocation)
    make[5]: *** [CMakeFiles/cmTC_5212d.dir/build.make:100: cmTC_5212d] Error 1
    make[5]: Leaving directory '/usr/local/tmp/crew/ponyc.20240622074005.dir/build/build_libs/gbenchmark-prefix/src/gbenchmark-build/CMakeFiles/CMakeScratch/TryCompile-3yBLsN'
    make[4]: *** [Makefile:127: cmTC_5212d/fast] Error 2
    make[4]: Leaving directory '/usr/local/tmp/crew/ponyc.20240622074005.dir/build/build_libs/gbenchmark-prefix/src/gbenchmark-build/CMakeFiles/CMakeScratch/TryCompile-3yBLsN'
    
    

  

  CMake will not be able to correctly generate this project.
Call Stack (most recent call first):
  CMakeLists.txt:17 (project)
```

--- a/bin/crew
+++ b/bin/crew
@@ -1284,13 +1284,10 @@ def install
   end
 
   # Add to installed packages list in devices.json, but remove first if it is already there.
-  if PackageUtils.installed?(@pkg.name)
-    crewlog "Removing package #{@pkg.name} from device.json since a version is already marked as installed."
-    @device[:installed_packages].delete_if { |entry| entry[:name] == @pkg.name }
-  end
   crewlog "Adding package #{@pkg.name} to device.json."
-  @device[:installed_packages].push(name: @pkg.name, version: @pkg.version, sha256: PackageUtils.get_sha256(@pkg, build_from_source: @opt_source))
+  @device[:installed_packages].delete_if { |entry| entry[:name] == @pkg.name } and @device[:installed_packages].push(name: @pkg.name, version: @pkg.version, sha256: PackageUtils.get_sha256(@pkg, build_from_source: @opt_source))
   save_json(@device)
+  crewlog "#{@pkg.name} in device.json after install: #{`jq --arg key '#{@pkg.name}' -e '.installed_packages[] | select(.name == $key )' #{File.join(CREW_CONFIG_PATH, 'device.json')}`}" if Kernel.system 'which jq', %i[out err] => File::NULL
 end
 
 def resolve_dependencies_and_build

--- a/lib/const.rb
+++ b/lib/const.rb
@@ -176,7 +176,7 @@ when 'x86_64'
 end
 
 CREW_LINKER = ENV.fetch('CREW_LINKER', 'mold')
-CREW_GLIBC_OVERRIDE_LINKER_FLAGS = (ARCH == 'x86_64' && LIBC_VERSION.to_f >= 2.35) ? " #{File.join(CREW_LIB_PREFIX, 'libC.so.6')} " : ''
+CREW_GLIBC_OVERRIDE_LINKER_FLAGS = ARCH == 'x86_64' && LIBC_VERSION.to_f >= 2.35 ? " #{File.join(CREW_LIB_PREFIX, 'libC.so.6')} " : ''
 CREW_LINKER_FLAGS = ENV.fetch('CREW_LINKER_FLAGS', CREW_GLIBC_OVERRIDE_LINKER_FLAGS)
 
 CREW_CORE_FLAGS           = "-O2 -pipe -ffat-lto-objects -fPIC #{CREW_ARCH_FLAGS} -fuse-ld=#{CREW_LINKER} #{CREW_LINKER_FLAGS}"

--- a/lib/const.rb
+++ b/lib/const.rb
@@ -2,7 +2,7 @@
 # Defines common constants used in different parts of crew
 require 'etc'
 
-CREW_VERSION = '1.49.13'
+CREW_VERSION = '1.50.0'
 
 # Kernel architecture.
 KERN_ARCH = Etc.uname[:machine]
@@ -175,8 +175,9 @@ when 'x86_64'
   CREW_ARCH_FLAGS = CREW_ARCH_FLAGS_OVERRIDE.to_s.empty? ? '' : CREW_ARCH_FLAGS_OVERRIDE
 end
 
-CREW_LINKER       = ENV.fetch('CREW_LINKER', 'mold')
-CREW_LINKER_FLAGS = ENV.fetch('CREW_LINKER_FLAGS', '')
+CREW_LINKER = ENV.fetch('CREW_LINKER', 'mold')
+CREW_GLIBC_OVERRIDE_LINKER_FLAGS = (ARCH == 'x86_64' && LIBC_VERSION.to_f >= 2.35) ? " #{File.join(CREW_LIB_PREFIX, 'libC.so.6')} " : ''
+CREW_LINKER_FLAGS = ENV.fetch('CREW_LINKER_FLAGS', CREW_GLIBC_OVERRIDE_LINKER_FLAGS)
 
 CREW_CORE_FLAGS           = "-O2 -pipe -ffat-lto-objects -fPIC #{CREW_ARCH_FLAGS} -fuse-ld=#{CREW_LINKER} #{CREW_LINKER_FLAGS}"
 CREW_COMMON_FLAGS         = "#{CREW_CORE_FLAGS} -flto=auto"

--- a/lib/fixup.rb
+++ b/lib/fixup.rb
@@ -196,15 +196,15 @@ if (ARCH == 'x86_64') && (LIBC_VERSION.to_f >= 2.36)
   # Link the system libc.so.6 to also require our renamed libC.so.6
   # which provides the float128 functions strtof128, strfromf128,
   # and __strtof128_nan.
-  @libc_patch_libraries = %w[libstdc++.so.6]
-  @libc_patch_libraries.delete_if { |lib| !File.file?(File.join(CREW_LIB_PREFIX, lib)) }
-  @libc_patch_libraries.delete_if { |lib| Kernel.system "patchelf --print-needed #{File.join(CREW_LIB_PREFIX, lib)} | grep -q libC.so.6" }
+  libc_patch_libraries = %w[libstdc++.so.6]
+  libc_patch_libraries.delete_if { |lib| !File.file?(File.join(CREW_LIB_PREFIX, lib)) }
+  libc_patch_libraries.delete_if { |lib| Kernel.system "patchelf --print-needed #{File.join(CREW_LIB_PREFIX, lib)} | grep -q libC.so.6" }
 
-  return if @libc_patch_libraries.empty?
+  return if libc_patch_libraries.empty?
 
   if File.file?(File.join(CREW_LIB_PREFIX, 'libC.so.6'))
     Dir.chdir(CREW_LIB_PREFIX) do
-      @libc_patch_libraries.each do |lib|
+      libc_patch_libraries.each do |lib|
         Kernel.system "patchelf --add-needed libC.so.6 #{lib}" and Kernel.system "patchelf --remove-needed libc.so.6 #{lib}"
         puts "#{lib} patched for use with Chromebrew's glibc.".lightgreen
       end

--- a/lib/fixup.rb
+++ b/lib/fixup.rb
@@ -189,3 +189,27 @@ end
 
 # Remove pagerenv tmp file in CREW_PACKAGES_PATH if it exists
 FileUtils.rm "#{CREW_PACKAGES_PATH}/pagerenv" if File.file?("#{CREW_PACKAGES_PATH}/pagerenv")
+
+# Handle broken system glibc affecting gcc_lib on newer x86_64 ChromeOS milestones.
+if (ARCH == 'x86_64') && (LIBC_VERSION.to_f >= 2.36)
+  abort("patchelf is needed. Please run: 'crew install patchelf && crew update'") unless File.file?(File.join(CREW_PREFIX, 'bin/patchelf'))
+  # Link the system libc.so.6 to also require our renamed libC.so.6
+  # which provides the float128 functions strtof128, strfromf128,
+  # and __strtof128_nan.
+  @libc_patch_libraries = %w[libstdc++.so.6]
+  @libc_patch_libraries.delete_if { |lib| !File.file?(File.join(CREW_LIB_PREFIX, lib)) }
+  @libc_patch_libraries.delete_if { |lib| Kernel.system "patchelf --print-needed #{File.join(CREW_LIB_PREFIX, lib)} | grep -q libC.so.6" }
+
+  return if @libc_patch_libraries.empty?
+
+  if File.file?(File.join(CREW_LIB_PREFIX, 'libC.so.6'))
+    Dir.chdir(CREW_LIB_PREFIX) do
+      @libc_patch_libraries.each do |lib|
+        Kernel.system "patchelf --add-needed libC.so.6 #{lib}" and Kernel.system "patchelf --remove-needed libc.so.6 #{lib}"
+        puts "#{lib} patched for use with Chromebrew's glibc.".lightgreen
+      end
+    end
+  else
+    puts 'The Chromebrew libC.so.6 was not found. Please reinstall glibc with this command: yes | crew reinstall glibc_lib237 && yes | crew upgrade'.lightred
+  end
+end

--- a/lib/fixup.rb
+++ b/lib/fixup.rb
@@ -209,7 +209,5 @@ if (ARCH == 'x86_64') && (LIBC_VERSION.to_f >= 2.36)
         puts "#{lib} patched for use with Chromebrew's glibc.".lightgreen
       end
     end
-  else
-    puts 'The Chromebrew libC.so.6 was not found. Please reinstall glibc with this command: yes | crew reinstall glibc_lib237 && yes | crew upgrade'.lightred
   end
 end

--- a/manifest/armv7l/g/glibc_build237.filelist
+++ b/manifest/armv7l/g/glibc_build237.filelist
@@ -769,7 +769,6 @@
 /usr/local/lib/libm.so.6
 /usr/local/lib/libmcheck.a
 /usr/local/lib/libmemusage.so
-/usr/local/lib/libnsl.so.2
 /usr/local/lib/libnss_compat.so
 /usr/local/lib/libnss_compat.so.2
 /usr/local/lib/libnss_db.so

--- a/manifest/armv7l/g/glibc_lib237.filelist
+++ b/manifest/armv7l/g/glibc_lib237.filelist
@@ -288,7 +288,6 @@
 /usr/local/lib/libm.so.6
 /usr/local/lib/libmemusage.so
 /usr/local/lib/libmvec.so.1
-/usr/local/lib/libnsl.so.2
 /usr/local/lib/libnss_compat.so
 /usr/local/lib/libnss_compat.so.2
 /usr/local/lib/libnss_db.so

--- a/manifest/x86_64/g/glibc_build237.filelist
+++ b/manifest/x86_64/g/glibc_build237.filelist
@@ -757,6 +757,7 @@
 /usr/local/lib64/libBrokenLocale.a
 /usr/local/lib64/libBrokenLocale.so
 /usr/local/lib64/libBrokenLocale.so.1
+/usr/local/lib64/libC.so.6
 /usr/local/lib64/libanl.a
 /usr/local/lib64/libanl.so
 /usr/local/lib64/libanl.so.1
@@ -769,9 +770,6 @@
 /usr/local/lib64/libcrypt.a
 /usr/local/lib64/libcrypt.so
 /usr/local/lib64/libcrypt.so.1
-/usr/local/lib64/libcrypt.so.1.1.0
-/usr/local/lib64/libcrypt.so.2
-/usr/local/lib64/libcrypt.so.2.0.0
 /usr/local/lib64/libdl.a
 /usr/local/lib64/libdl.so.2
 /usr/local/lib64/libg.a
@@ -784,7 +782,6 @@
 /usr/local/lib64/libmvec.a
 /usr/local/lib64/libmvec.so
 /usr/local/lib64/libmvec.so.1
-/usr/local/lib64/libnsl.so.2
 /usr/local/lib64/libnss_compat.so
 /usr/local/lib64/libnss_compat.so.2
 /usr/local/lib64/libnss_db.so

--- a/manifest/x86_64/g/glibc_lib237.filelist
+++ b/manifest/x86_64/g/glibc_lib237.filelist
@@ -275,6 +275,7 @@
 /usr/local/lib64/ld-linux-x86-64.so.2
 /usr/local/lib64/libBrokenLocale.so
 /usr/local/lib64/libBrokenLocale.so.1
+/usr/local/lib64/libC.so.6
 /usr/local/lib64/libanl.so
 /usr/local/lib64/libanl.so.1
 /usr/local/lib64/libc.so
@@ -283,16 +284,12 @@
 /usr/local/lib64/libc_malloc_debug.so.0
 /usr/local/lib64/libcrypt.so
 /usr/local/lib64/libcrypt.so.1
-/usr/local/lib64/libcrypt.so.1.1.0
-/usr/local/lib64/libcrypt.so.2
-/usr/local/lib64/libcrypt.so.2.0.0
 /usr/local/lib64/libdl.so.2
 /usr/local/lib64/libm.so
 /usr/local/lib64/libm.so.6
 /usr/local/lib64/libmemusage.so
 /usr/local/lib64/libmvec.so
 /usr/local/lib64/libmvec.so.1
-/usr/local/lib64/libnsl.so.2
 /usr/local/lib64/libnss_compat.so
 /usr/local/lib64/libnss_compat.so.2
 /usr/local/lib64/libnss_db.so

--- a/packages/glibc_build237.rb
+++ b/packages/glibc_build237.rb
@@ -243,8 +243,8 @@ class Glibc_build237 < Package
           # Link our libm to also require our renamed libC.so.6
           # which provides the float128 functions strtof128, strfromf128,
           # and __strtof128_nan.
-          @libc_patch_libraries = %w[libm.so.6]
-          @libc_patch_libraries.each do |lib|
+          libc_patch_libraries = %w[libm.so.6]
+          libc_patch_libraries.each do |lib|
             system "patchelf --replace-needed libc.so.6 libC.so.6 #{lib}"
           end
         end
@@ -274,15 +274,15 @@ class Glibc_build237 < Package
       # Link the system libc.so.6 to also require our renamed libC.so.6
       # which provides the float128 functions strtof128, strfromf128,
       # and __strtof128_nan.
-      @libc_patch_libraries = %w[libc.so.6 libm.so.6 libstdc++.so.6]
-      @libc_patch_libraries.delete_if { |lib| !File.file?(File.join(CREW_LIB_PREFIX, lib)) }
-      @libc_patch_libraries.delete_if { |lib| Kernel.system "patchelf --print-needed #{File.join(CREW_LIB_PREFIX, lib)} | grep -q libC.so.6" }
+      libc_patch_libraries = %w[libc.so.6 libm.so.6 libstdc++.so.6]
+      libc_patch_libraries.delete_if { |lib| !File.file?(File.join(CREW_LIB_PREFIX, lib)) }
+      libc_patch_libraries.delete_if { |lib| Kernel.system "patchelf --print-needed #{File.join(CREW_LIB_PREFIX, lib)} | grep -q libC.so.6" }
 
-      return if @libc_patch_libraries.empty?
+      return if libc_patch_libraries.empty?
 
       if File.file?(File.join(CREW_LIB_PREFIX, 'libC.so.6'))
         Dir.chdir(CREW_LIB_PREFIX) do
-          @libc_patch_libraries.each do |lib|
+          libc_patch_libraries.each do |lib|
             Kernel.system "patchelf --add-needed libC.so.6 #{lib}" and Kernel.system "patchelf --remove-needed libc.so.6 #{lib}"
             puts "#{lib} patched for use with Chromebrew's glibc.".lightgreen
           end

--- a/packages/glibc_build237.rb
+++ b/packages/glibc_build237.rb
@@ -252,7 +252,7 @@ class Glibc_build237 < Package
     end
     # Only save libnsl.so.2, since libnsl.so.1 is provided by perl
     # For this to work, build on a M107 or newer container.
-    FileUtils.cp File.realpath("#{CREW_DEST_LIB_PREFIX}/libnsl.so.1"), "#{CREW_DEST_LIB_PREFIX}/libnsl.so.2" if LIBC_VERSION >= 2.35
+    FileUtils.cp File.realpath("#{CREW_DEST_LIB_PREFIX}/libnsl.so.1"), "#{CREW_DEST_LIB_PREFIX}/libnsl.so.2" if LIBC_VERSION.to_f >= 2.35
     FileUtils.rm_f "#{CREW_DEST_LIB_PREFIX}/libnsl.so"
     FileUtils.rm_f "#{CREW_DEST_LIB_PREFIX}/libnsl.so.1"
 
@@ -268,7 +268,7 @@ class Glibc_build237 < Package
 
   def self.postinstall
     # Handle broken system glibc affecting system glibc and libm.so.6 on newer x86_64 ChromeOS milestones.
-    if (ARCH == 'x86_64') && (LIBC_VERSION >= 2.35)
+    if (ARCH == 'x86_64') && (LIBC_VERSION.to_f >= 2.35)
       FileUtils.cp "/#{ARCH_LIB}/libc.so.6", File.join(CREW_LIB_PREFIX, 'libc.so.6.system') and FileUtils.mv File.join(CREW_LIB_PREFIX, 'libc.so.6.system'), File.join(CREW_LIB_PREFIX, 'libc.so.6')
       puts "patchelf is needed. Please run: 'crew install patchelf ; crew postinstall #{name}'".lightred unless File.file?(File.join(CREW_PREFIX, 'bin/patchelf'))
       # Link the system libc.so.6 to also require our renamed libC.so.6

--- a/packages/glibc_build237.rb
+++ b/packages/glibc_build237.rb
@@ -270,7 +270,7 @@ class Glibc_build237 < Package
     # Handle broken system glibc affecting system glibc and libm.so.6 on newer x86_64 ChromeOS milestones.
     if (ARCH == 'x86_64') && (LIBC_VERSION.to_f >= 2.35)
       FileUtils.cp "/#{ARCH_LIB}/libc.so.6", File.join(CREW_LIB_PREFIX, 'libc.so.6.system') and FileUtils.mv File.join(CREW_LIB_PREFIX, 'libc.so.6.system'), File.join(CREW_LIB_PREFIX, 'libc.so.6')
-      abort("patchelf is needed. Please run: 'crew install patchelf ; crew postinstall #{name}'") unless File.file?(File.join(CREW_PREFIX, 'bin/patchelf'))
+      puts "patchelf is needed. Please run: 'crew install patchelf ; crew postinstall #{name}'".lightred unless File.file?(File.join(CREW_PREFIX, 'bin/patchelf'))
       # Link the system libc.so.6 to also require our renamed libC.so.6
       # which provides the float128 functions strtof128, strfromf128,
       # and __strtof128_nan.
@@ -287,8 +287,6 @@ class Glibc_build237 < Package
             puts "#{lib} patched for use with Chromebrew's glibc.".lightgreen
           end
         end
-      else
-        abort('The Chromebrew libC.so.6 was not found. Please reinstall glibc.')
       end
     end
 

--- a/packages/glibc_build237.rb
+++ b/packages/glibc_build237.rb
@@ -6,21 +6,23 @@ class Glibc_build237 < Package
   license 'LGPL-2.1+, BSD, HPND, ISC, inner-net, rc, and PCRE'
   compatibility 'all'
   binary_compression 'tar.zst'
-  @libc_version = LIBC_VERSION
-  version '2.37'
+  # @libc_version = LIBC_VERSION
+  @libc_version = '2.37'
+  version '2.37-1'
   source_url 'https://github.com/bminor/glibc.git'
   git_hashtag 'glibc-2.37'
 
   binary_sha256({
-    aarch64: 'd9cac82b17d463b98caf5c29e143775d3aeed29df3851207ad930d9c5b4c391b',
-     armv7l: 'd9cac82b17d463b98caf5c29e143775d3aeed29df3851207ad930d9c5b4c391b',
-     x86_64: '07754223434ad59930ebca10f0adbd390700ebd93a43853b183ed30f0a0a33ca'
+    aarch64: 'fffbe32d69e3d972c9c845e8fa244159d7ac8b82ae9f07206c7189b20e3457c8',
+     armv7l: 'fffbe32d69e3d972c9c845e8fa244159d7ac8b82ae9f07206c7189b20e3457c8',
+     x86_64: '890a2ba75346498bb20165d55ee2d91bcb57d8c65df485029a5b8a20a5f73c19'
   })
 
   depends_on 'gawk' => :build
   depends_on 'filecmd' # L Fixes creating symlinks on a fresh install.
   depends_on 'libidn2' => :build
   depends_on 'texinfo' => :build
+  depends_on 'patchelf' # L
 
   conflicts_ok
   no_env_options
@@ -37,8 +39,21 @@ class Glibc_build237 < Package
     system "sed -i 's,verbose,locale_verbose,g' fedora/build-locale-archive.c"
     system "sed -i 's,be_quiet,locale_be_quiet,g' fedora/build-locale-archive.c"
 
-    @googlesource_branch = 'release-R123-15786.B'
+    @googlesource_branch = 'release-R128-15964.B'
     system "git clone --depth=1 -b  #{@googlesource_branch} https://chromium.googlesource.com/chromiumos/overlays/chromiumos-overlay googlesource"
+    # We need to avoid this patch to preserve float128. Otherwise strtof128,
+    # strfromf128, and __strtof128_nan, are not available in our glibc.
+    # This still requires hackery to avoid issues, as handled below by
+    # renaming our libc.so.6 to libC.so.6 and using patchelf to add
+    # needs for our libm.so.6 as well as the system provided libc.so.6.
+    # gcc_lib's stdc++.so.6 also needs to be told to need our glibc's
+    # renamed libC.so.6. Additionally, packages on these systems need to
+    # be built with CREW_LINKER_FLAGS=' /usr/local/lib64/libC.so.6 ' set.
+    #
+    FileUtils.rm 'googlesource/sys-libs/glibc/files/local/glibc-2.37/0003-Disable-float128-support-for-x86_64-x86.patch'
+    # Please submit suggestions for less hacky workarounds.
+    #
+
     Dir.glob('googlesource/sys-libs/glibc/files/local/glibc-2.37/*.patch').each do |patch|
       puts "patch -Np1 < #{patch} || true" if @opt_verbose
       system "patch -Np1 -F 10 -i #{patch} || true"
@@ -98,10 +113,11 @@ class Glibc_build237 < Package
           system "sed -i 's,install-symbolic-link,/bin/true,g' ../Makefile"
           system "sed -i 's,symbolic-link-prog := $(elf-objpfx)sln,symbolic-link-prog := /bin/true,g' ../Makerules"
         when 'x86_64'
+          # https://source.chromium.org/chromiumos/_/chromium/chromiumos/overlays/chromiumos-overlay/+/a73162d56fd689b26812282c3f9bc4f2b5a67530:sys-libs/glibc/glibc-2.37.ebuild
           File.write('configparms', "slibdir=#{CREW_LIB_PREFIX}", mode: 'a+')
           system "CFLAGS='-fuse-ld=mold -pipe -O2 -fipa-pta \
           -fno-semantic-interposition -falign-functions=32 \
-          -fdevirtualize-at-ltrans' \
+          -fdevirtualize-at-ltrans -mstackrealign' \
             ../configure \
             --prefix=#{CREW_PREFIX} \
             --libdir=#{CREW_LIB_PREFIX} \
@@ -171,52 +187,72 @@ class Glibc_build237 < Package
       end
       if @libc_version.to_f >= 2.32
         system "install -Dt #{CREW_DEST_PREFIX}/bin -m755 build-locale-archive"
-        system "make -j1 DESTDIR=#{CREW_DEST_DIR} localedata/install-locales"
+        system "make -j1 DESTDIR=#{CREW_DEST_DIR} localedata/install-locales || make -j1 DESTDIR=#{CREW_DEST_DIR} localedata/install-locales"
       end
+
       Dir.chdir CREW_DEST_LIB_PREFIX do
         puts "System glibc version is #{LIBC_VERSION}.".lightblue
         puts 'Creating symlinks to system glibc version to prevent breakage.'.lightblue
         case ARCH
         when 'aarch64', 'armv7l'
-          FileUtils.ln_sf "/lib/ld-#{@libc_version}.so", 'ld-linux-armhf.so.3'
+          FileUtils.ln_sf '/lib/ld-linux-armhf.so.3', 'ld-linux-armhf.so.3'
         when 'i686'
           FileUtils.ln_sf "/lib/ld-#{@libc_version}.so", 'ld-linux-i686.so.2'
-        when 'x86_64'
-          FileUtils.ln_sf "/lib64/ld-#{@libc_version}.so", 'ld-linux-x86-64.so.2'
+          # newer x86_64 ChromeOS glibc lacks strtof128, strfromf128
+          # and __strtof128_nan
+          # when 'x86_64'
+          #   FileUtils.ln_sf '/lib64/ld-linux-x86-64.so.2', 'ld-linux-x86-64.so.2.system'
         end
-        @libraries = %w[ld libBrokenLocale libSegFault libanl libc libcrypt
-                        libdl libm libmemusage libmvec libnsl libnss_compat libnss_db
-                        libnss_dns libnss_files libnss_hesiod libpcprofile libpthread
-                        libthread_db libresolv librlv librt libthread_db-1.0 libutil]
-        @libraries -= ['libpthread'] if @libc_version.to_f >= 2.35
-        @libraries.each do |lib|
-          # Reject entries which aren't libraries ending in .so, and which aren't files.
-          Dir["/#{ARCH_LIB}/#{lib}.so*"].reject { |f| File.directory?(f) }.each do |f|
-            @filetype = `file #{f}`.chomp
-            if ['shared object', 'symbolic link'].any? { |type| @filetype.include?(type) }
-              g = File.basename(f)
-              FileUtils.ln_sf f.to_s, "#{CREW_DEST_LIB_PREFIX}/#{g}"
-            end
-          end
-          # Reject entries which aren't libraries ending in .so, and which aren't files.
-          # Reject text files such as libc.so because they points to files like
-          # libc_nonshared.a, which are not provided by ChromeOS
-          Dir["/usr/#{ARCH_LIB}/#{lib}.so*"].reject { |f| File.directory?(f) }.each do |f|
-            @filetype = `file #{f}`.chomp
-            puts "f: #{@filetype}" if @opt_verbose
-            if ['shared object', 'symbolic link'].any? { |type| @filetype.include?(type) }
-              g = File.basename(f)
-              FileUtils.ln_sf f.to_s, "#{CREW_DEST_LIB_PREFIX}/#{g}"
-            elsif @opt_verbose
-              puts "#{f} excluded because #{@filetype}"
-            end
+        if ARCH == 'x86_64'
+          # Save our copy of libc.so.6
+          FileUtils.mv File.join(CREW_DEST_LIB_PREFIX, 'libc.so.6'), File.join(CREW_DEST_LIB_PREFIX, 'libC.so.6')
+
+          # Make a symlink to the system libc.so.6, which will require patchelf run on it in the postinstall.
+          FileUtils.ln_sf "/#{ARCH_LIB}/libc.so.6", 'libc.so.6'
+
+          ## Also save our copy of libm.so.6 since it has the float128 functions.
+          # @libraries = %w[ld libBrokenLocale libSegFault libanl libc libcrypt
+          # libdl libmemusage libmvec libnsl libnss_compat libnss_db
+          # libnss_dns libnss_files libnss_hesiod libpcprofile libpthread
+          # libthread_db libresolv librlv librt libthread_db-1.0 libutil]
+          # @libraries -= ['libpthread'] if @libc_version.to_f >= 2.35
+          # @libraries.each do |lib|
+          ## Reject entries which aren't libraries ending in .so, and which aren't files.
+          # Dir["/#{ARCH_LIB}/#{lib}.so*"].reject { |f| File.directory?(f) }.each do |f|
+          # @filetype = `file #{f}`.chomp
+          # next unless ['shared object', 'symbolic link'].any? { |type| @filetype.include?(type) }
+          # g = File.basename(f)
+          # FileUtils.mkdir_p CREW_DEST_LIB_PREFIX
+          # Dir.chdir(CREW_DEST_LIB_PREFIX) do
+          # FileUtils.ln_sf f.to_s, g.to_s
+          # end
+          # end
+          ## Reject entries which aren't libraries ending in .so, and which aren't files.
+          ## Reject text files such as libc.so because they points to files like
+          ## libc_nonshared.a, which are not provided by ChromeOS
+          # Dir["/usr/#{ARCH_LIB}/#{lib}.so*"].reject { |f| File.directory?(f) }.each do |f|
+          # @filetype = `file #{f}`.chomp
+          # puts "f: #{@filetype}" if @opt_verbose
+          # if ['shared object', 'symbolic link'].any? { |type| @filetype.include?(type) }
+          # g = File.basename(f)
+          # FileUtils.ln_sf f.to_s, "#{CREW_DEST_LIB_PREFIX}/#{g}"
+          # elsif @opt_verbose
+          # puts "#{f} excluded because #{@filetype}"
+          # end
+          # end
+          # Link our libm to also require our renamed libC.so.6
+          # which provides the float128 functions strtof128, strfromf128,
+          # and __strtof128_nan.
+          @libc_patch_libraries = %w[libm.so.6]
+          @libc_patch_libraries.each do |lib|
+            system "patchelf --replace-needed libc.so.6 libC.so.6 #{lib}"
           end
         end
       end
     end
     # Only save libnsl.so.2, since libnsl.so.1 is provided by perl
     # For this to work, build on a M107 or newer container.
-    FileUtils.cp File.realpath("#{CREW_DEST_LIB_PREFIX}/libnsl.so.1"), "#{CREW_DEST_LIB_PREFIX}/libnsl.so.2"
+    FileUtils.cp File.realpath("#{CREW_DEST_LIB_PREFIX}/libnsl.so.1"), "#{CREW_DEST_LIB_PREFIX}/libnsl.so.2" if LIBC_VERSION.to_f >= 2.35
     FileUtils.rm_f "#{CREW_DEST_LIB_PREFIX}/libnsl.so"
     FileUtils.rm_f "#{CREW_DEST_LIB_PREFIX}/libnsl.so.1"
 
@@ -231,49 +267,83 @@ class Glibc_build237 < Package
   end
 
   def self.postinstall
-    if File.exist?("#{CREW_LIB_PREFIX}/libc.so.6")
-      @crew_libcvertokens = `#{CREW_LIB_PREFIX}/libc.so.6`.lines.first.chomp.split(/\s/)
-      @libc_version = @crew_libcvertokens[@crew_libcvertokens.find_index('version') + 1].sub!(/[[:punct:]]?$/, '')
-      puts "Package glibc version is #{@libc_version}.".lightblue
-    else
-      @libc_version = LIBC_VERSION
-    end
-    @libraries = %w[ld libBrokenLocale libSegFault libanl libc libcrypt
-                    libdl libm libmemusage libmvec libnsl libnss_compat libnss_db
-                    libnss_dns libnss_files libnss_hesiod libpcprofile libpthread
-                    libthread_db libresolv librlv librt libthread_db-1.0 libutil]
-    @libraries -= ['libpthread'] if @libc_version.to_f >= 2.35
-    Dir.chdir CREW_LIB_PREFIX do
-      puts "System glibc version is #{@libc_version}.".lightblue
-      puts 'Creating symlinks to system glibc version to prevent breakage.'.lightblue
-      case ARCH
-      when 'aarch64', 'armv7l'
-        FileUtils.ln_sf '/lib/ld-linux-armhf.so.3', 'ld-linux-armhf.so.3'
-      when 'i686'
-        FileUtils.ln_sf "/lib/ld-#{@libc_version}.so", 'ld-linux-i686.so.2'
-      when 'x86_64'
-        FileUtils.ln_sf '/lib64/ld-linux-x86-64.so.2', 'ld-linux-x86-64.so.2'
-      end
-      @libraries.each do |lib|
-        # Reject entries which aren't libraries ending in .so, and which aren't files.
-        Dir["/#{ARCH_LIB}/#{lib}.so*"].reject { |f| File.directory?(f) }.each do |f|
-          @filetype = `file #{f}`.chomp
-          if ['shared object', 'symbolic link'].any? { |type| @filetype.include?(type) }
-            g = File.basename(f)
-            FileUtils.ln_sf f.to_s, "#{CREW_LIB_PREFIX}/#{g}"
+    # Handle broken system glibc affecting system glibc and libm.so.6 on newer x86_64 ChromeOS milestones.
+    if (ARCH == 'x86_64') && (LIBC_VERSION.to_f >= 2.35)
+      FileUtils.cp "/#{ARCH_LIB}/libc.so.6", File.join(CREW_LIB_PREFIX, 'libc.so.6.system') and FileUtils.mv File.join(CREW_LIB_PREFIX, 'libc.so.6.system'), File.join(CREW_LIB_PREFIX, 'libc.so.6')
+      abort("patchelf is needed. Please run: 'crew install patchelf ; crew postinstall #{name}'") unless File.file?(File.join(CREW_PREFIX, 'bin/patchelf'))
+      # Link the system libc.so.6 to also require our renamed libC.so.6
+      # which provides the float128 functions strtof128, strfromf128,
+      # and __strtof128_nan.
+      @libc_patch_libraries = %w[libc.so.6 libm.so.6]
+      @libc_patch_libraries.delete_if { |lib| !File.file?(File.join(CREW_LIB_PREFIX, lib)) }
+      @libc_patch_libraries.delete_if { |lib| Kernel.system "patchelf --print-needed #{File.join(CREW_LIB_PREFIX, lib)} | grep -q libC.so.6" }
+
+      return if @libc_patch_libraries.empty?
+
+      if File.file?(File.join(CREW_LIB_PREFIX, 'libC.so.6'))
+        Dir.chdir(CREW_LIB_PREFIX) do
+          @libc_patch_libraries.each do |lib|
+            Kernel.system "patchelf --add-needed libC.so.6 #{lib}" and Kernel.system "patchelf --remove-needed libc.so.6 #{lib}"
+            puts "#{lib} patched for use with Chromebrew's glibc.".lightgreen
           end
         end
-        # Reject entries which aren't libraries ending in .so, and which aren't files.
-        # Reject text files such as libc.so because they points to files like
-        # libc_nonshared.a, which are not provided by ChromeOS
-        Dir["/usr/#{ARCH_LIB}/#{lib}.so*"].reject { |f| File.directory?(f) }.each do |f|
-          @filetype = `file #{f}`.chomp
-          puts "f: #{@filetype}" if @opt_verbose
-          if ['shared object', 'symbolic link'].any? { |type| @filetype.include?(type) }
+      else
+        abort('The Chromebrew libC.so.6 was not found. Please reinstall glibc.')
+      end
+    end
+
+    unless ARCH == 'x86_64'
+      if File.exist?("#{CREW_LIB_PREFIX}/libc.so.6")
+        @crew_libcvertokens = `#{CREW_LIB_PREFIX}/libc.so.6`.lines.first.chomp.split(/\s/)
+        @libc_version = @crew_libcvertokens[@crew_libcvertokens.find_index('version') + 1].sub!(/[[:punct:]]?$/, '')
+        puts "Package glibc version is #{@libc_version}.".lightblue
+      else
+        @libc_version = LIBC_VERSION
+      end
+      @libraries = %w[ld libBrokenLocale libSegFault libanl libc libcrypt
+                      libdl libm libmemusage libmvec libnsl libnss_compat libnss_db
+                      libnss_dns libnss_files libnss_hesiod libpcprofile libpthread
+                      libthread_db libresolv librlv librt libthread_db-1.0 libutil]
+      @libraries -= ['libpthread'] if @libc_version.to_f >= 2.35
+      @libraries -= ['libc'] if (@libc_version.to_f >= 2.35) && Kernel.system("patchelf --print-needed #{File.join(CREW_LIB_PREFIX, 'libc.so.6')} | grep -q libC.so.6")
+      @libraries -= ['libm'] if (@libc_version.to_f >= 2.35) && Kernel.system("patchelf --print-needed #{File.join(CREW_LIB_PREFIX, 'libm.so.6')} | grep -q libC.so.6")
+      Dir.chdir CREW_LIB_PREFIX do
+        puts "System glibc version is #{@libc_version}.".lightblue
+        puts 'Creating symlinks to system glibc version to prevent breakage.'.lightblue
+        case ARCH
+        when 'aarch64', 'armv7l'
+          FileUtils.ln_sf '/lib/ld-linux-armhf.so.3', 'ld-linux-armhf.so.3'
+        when 'i686'
+          FileUtils.ln_sf "/lib/ld-#{@libc_version}.so", 'ld-linux-i686.so.2'
+          # newer x86_64 ChromeOS glibc lacks strtof128, strfromf128
+          # and __strtof128_nan
+          # when 'x86_64'
+          #   FileUtils.ln_sf '/lib64/ld-linux-x86-64.so.2', 'ld-linux-x86-64.so.2'
+        end
+        @libraries.each do |lib|
+          # Reject entries which aren't libraries ending in .so, and which aren't files.
+          Dir["/#{ARCH_LIB}/#{lib}.so*"].reject { |f| File.directory?(f) }.each do |f|
+            @filetype = `file #{f}`.chomp
+            next unless ['shared object', 'symbolic link'].any? { |type| @filetype.include?(type) }
             g = File.basename(f)
+            next if Kernel.system "patchelf --print-needed #{File.join(CREW_LIB_PREFIX, g)} | grep -q libC.so.6"
+
             FileUtils.ln_sf f.to_s, "#{CREW_LIB_PREFIX}/#{g}"
-          elsif @opt_verbose
-            puts "#{f} excluded because #{@filetype}"
+          end
+          # Reject entries which aren't libraries ending in .so, and which aren't files.
+          # Reject text files such as libc.so because they points to files like
+          # libc_nonshared.a, which are not provided by ChromeOS
+          Dir["/usr/#{ARCH_LIB}/#{lib}.so*"].reject { |f| File.directory?(f) }.each do |f|
+            @filetype = `file #{f}`.chomp
+            puts "f: #{@filetype}" if @opt_verbose
+            if ['shared object', 'symbolic link'].any? { |type| @filetype.include?(type) }
+              g = File.basename(f)
+              next if Kernel.system "patchelf --print-needed #{File.join(CREW_LIB_PREFIX, g)} | grep -q libC.so.6"
+
+              FileUtils.ln_sf f.to_s, "#{CREW_LIB_PREFIX}/#{g}"
+            elsif @opt_verbose
+              puts "#{f} excluded because #{@filetype}"
+            end
           end
         end
       end

--- a/packages/glibc_build237.rb
+++ b/packages/glibc_build237.rb
@@ -274,7 +274,7 @@ class Glibc_build237 < Package
       # Link the system libc.so.6 to also require our renamed libC.so.6
       # which provides the float128 functions strtof128, strfromf128,
       # and __strtof128_nan.
-      @libc_patch_libraries = %w[libc.so.6 libm.so.6]
+      @libc_patch_libraries = %w[libc.so.6 libm.so.6 libstdc++.so.6]
       @libc_patch_libraries.delete_if { |lib| !File.file?(File.join(CREW_LIB_PREFIX, lib)) }
       @libc_patch_libraries.delete_if { |lib| Kernel.system "patchelf --print-needed #{File.join(CREW_LIB_PREFIX, lib)} | grep -q libC.so.6" }
 

--- a/packages/glibc_build237.rb
+++ b/packages/glibc_build237.rb
@@ -252,7 +252,7 @@ class Glibc_build237 < Package
     end
     # Only save libnsl.so.2, since libnsl.so.1 is provided by perl
     # For this to work, build on a M107 or newer container.
-    FileUtils.cp File.realpath("#{CREW_DEST_LIB_PREFIX}/libnsl.so.1"), "#{CREW_DEST_LIB_PREFIX}/libnsl.so.2" if LIBC_VERSION.to_f >= 2.35
+    FileUtils.cp File.realpath("#{CREW_DEST_LIB_PREFIX}/libnsl.so.1"), "#{CREW_DEST_LIB_PREFIX}/libnsl.so.2" if LIBC_VERSION >= 2.35
     FileUtils.rm_f "#{CREW_DEST_LIB_PREFIX}/libnsl.so"
     FileUtils.rm_f "#{CREW_DEST_LIB_PREFIX}/libnsl.so.1"
 
@@ -268,14 +268,14 @@ class Glibc_build237 < Package
 
   def self.postinstall
     # Handle broken system glibc affecting system glibc and libm.so.6 on newer x86_64 ChromeOS milestones.
-    if (ARCH == 'x86_64') && (LIBC_VERSION.to_f >= 2.35)
+    if (ARCH == 'x86_64') && (LIBC_VERSION >= 2.35)
       FileUtils.cp "/#{ARCH_LIB}/libc.so.6", File.join(CREW_LIB_PREFIX, 'libc.so.6.system') and FileUtils.mv File.join(CREW_LIB_PREFIX, 'libc.so.6.system'), File.join(CREW_LIB_PREFIX, 'libc.so.6')
       puts "patchelf is needed. Please run: 'crew install patchelf ; crew postinstall #{name}'".lightred unless File.file?(File.join(CREW_PREFIX, 'bin/patchelf'))
       # Link the system libc.so.6 to also require our renamed libC.so.6
       # which provides the float128 functions strtof128, strfromf128,
       # and __strtof128_nan.
       libc_patch_libraries = %w[libc.so.6 libm.so.6 libstdc++.so.6]
-      libc_patch_libraries.delete_if { |lib| !File.file?(File.join(CREW_LIB_PREFIX, lib)) }
+      libc_patch_libraries.keep_if { |lib| File.file?(File.join(CREW_LIB_PREFIX, lib)) }
       libc_patch_libraries.delete_if { |lib| Kernel.system "patchelf --print-needed #{File.join(CREW_LIB_PREFIX, lib)} | grep -q libC.so.6" }
 
       return if libc_patch_libraries.empty?

--- a/packages/glibc_dev237.rb
+++ b/packages/glibc_dev237.rb
@@ -4,16 +4,16 @@ Package.load_package("#{__dir__}/glibc_build237.rb")
 class Glibc_dev237 < Package
   description 'glibc: everything except what is in glibc_lib'
   homepage Glibc_build237.homepage
-  version '2.37'
+  version '2.37-patchelf1'
   license Glibc_build237.license
   compatibility 'x86_64 aarch64 armv7l'
   source_url 'SKIP'
   binary_compression 'tar.zst'
 
   binary_sha256({
-    aarch64: '73b6371edbfd4295465c24dbc9fb60975a4c37aff37a6eecf940d22b2727d218',
-     armv7l: '73b6371edbfd4295465c24dbc9fb60975a4c37aff37a6eecf940d22b2727d218',
-     x86_64: '9ca2f35c8325766fb20cd4cd4c02e7b46052158ca528a0807b33a8624a050de8'
+    aarch64: 'fc071b97e0eaa0e1e9b6bdad4ab378112dbef849453fbf3bc4e10e22260b1841',
+     armv7l: 'fc071b97e0eaa0e1e9b6bdad4ab378112dbef849453fbf3bc4e10e22260b1841',
+     x86_64: '72b99a7f86687958639d0935d686260f50afc45e0209ff717bbd42e4ef9d7bfb'
   })
 
   depends_on 'glibc_build237' => :build

--- a/packages/glibc_lib237.rb
+++ b/packages/glibc_lib237.rb
@@ -57,7 +57,7 @@ class Glibc_lib237 < Package
   end
 
   def self.postinstall
-    if (ARCH == 'x86_64') && (LIBC_VERSION >= 2.35)
+    if (ARCH == 'x86_64') && (LIBC_VERSION.to_f >= 2.35)
       FileUtils.cp "/#{ARCH_LIB}/libc.so.6", File.join(CREW_LIB_PREFIX, 'libc.so.6.system') and FileUtils.mv File.join(CREW_LIB_PREFIX, 'libc.so.6.system'), File.join(CREW_LIB_PREFIX, 'libc.so.6') unless Kernel.system "patchelf --print-needed #{File.join(CREW_LIB_PREFIX, 'libc.so.6')} | grep -q libC.so.6"
       puts "patchelf is needed. Please run: 'crew install patchelf ; crew postinstall #{name}'".lightred unless File.file?(File.join(CREW_PREFIX, 'bin/patchelf'))
       # Link the system libc.so.6 to also require our renamed libC.so.6

--- a/packages/glibc_lib237.rb
+++ b/packages/glibc_lib237.rb
@@ -63,15 +63,15 @@ class Glibc_lib237 < Package
       # Link the system libc.so.6 to also require our renamed libC.so.6
       # which provides the float128 functions strtof128, strfromf128,
       # and __strtof128_nan.
-      @libc_patch_libraries = %w[libc.so.6 libm.so.6 libstdc++.so.6]
-      @libc_patch_libraries.delete_if { |lib| !File.file?(File.join(CREW_LIB_PREFIX, lib)) }
-      @libc_patch_libraries.delete_if { |lib| Kernel.system "patchelf --print-needed #{File.join(CREW_LIB_PREFIX, lib)} | grep -q libC.so.6" }
+      libc_patch_libraries = %w[libc.so.6 libm.so.6 libstdc++.so.6]
+      libc_patch_libraries.delete_if { |lib| !File.file?(File.join(CREW_LIB_PREFIX, lib)) }
+      libc_patch_libraries.delete_if { |lib| Kernel.system "patchelf --print-needed #{File.join(CREW_LIB_PREFIX, lib)} | grep -q libC.so.6" }
 
-      return if @libc_patch_libraries.empty?
+      return if libc_patch_libraries.empty?
 
       if File.file?(File.join(CREW_LIB_PREFIX, 'libC.so.6'))
         Dir.chdir(CREW_LIB_PREFIX) do
-          @libc_patch_libraries.each do |lib|
+          libc_patch_libraries.each do |lib|
             Kernel.system "patchelf --add-needed libC.so.6 #{lib}" and Kernel.system "patchelf --remove-needed libc.so.6 #{lib}"
             puts "#{lib} patched for use with Chromebrew's glibc.".lightgreen
           end

--- a/packages/glibc_lib237.rb
+++ b/packages/glibc_lib237.rb
@@ -55,4 +55,28 @@ class Glibc_lib237 < Package
       end
     end
   end
+
+  def self.postinstall
+    if (ARCH == 'x86_64') && (LIBC_VERSION.to_f >= 2.35)
+      FileUtils.cp "/#{ARCH_LIB}/libc.so.6", File.join(CREW_LIB_PREFIX, 'libc.so.6.system') and FileUtils.mv File.join(CREW_LIB_PREFIX, 'libc.so.6.system'), File.join(CREW_LIB_PREFIX, 'libc.so.6') unless Kernel.system "patchelf --print-needed #{File.join(CREW_LIB_PREFIX, 'libc.so.6')} | grep -q libC.so.6"
+      puts "patchelf is needed. Please run: 'crew install patchelf ; crew postinstall #{name}'".lightred unless File.file?(File.join(CREW_PREFIX, 'bin/patchelf'))
+      # Link the system libc.so.6 to also require our renamed libC.so.6
+      # which provides the float128 functions strtof128, strfromf128,
+      # and __strtof128_nan.
+      @libc_patch_libraries = %w[libc.so.6 libm.so.6 libstdc++.so.6]
+      @libc_patch_libraries.delete_if { |lib| !File.file?(File.join(CREW_LIB_PREFIX, lib)) }
+      @libc_patch_libraries.delete_if { |lib| Kernel.system "patchelf --print-needed #{File.join(CREW_LIB_PREFIX, lib)} | grep -q libC.so.6" }
+
+      return if @libc_patch_libraries.empty?
+
+      if File.file?(File.join(CREW_LIB_PREFIX, 'libC.so.6'))
+        Dir.chdir(CREW_LIB_PREFIX) do
+          @libc_patch_libraries.each do |lib|
+            Kernel.system "patchelf --add-needed libC.so.6 #{lib}" and Kernel.system "patchelf --remove-needed libc.so.6 #{lib}"
+            puts "#{lib} patched for use with Chromebrew's glibc.".lightgreen
+          end
+        end
+      end
+    end
+  end
 end

--- a/packages/glibc_lib237.rb
+++ b/packages/glibc_lib237.rb
@@ -4,20 +4,19 @@ Package.load_package("#{__dir__}/glibc_build237.rb")
 class Glibc_lib237 < Package
   description 'glibc libraries'
   homepage Glibc_build237.homepage
-  version '2.37-1' # Do not use @_ver here, it will break the installer.
+  version '2.37-patchelf1' # Do not use @_ver here, it will break the installer.
   license Glibc_build237.license
   compatibility 'x86_64 aarch64 armv7l'
   source_url 'SKIP'
   binary_compression 'tar.zst'
 
   binary_sha256({
-    aarch64: 'efb6e1f5593f97b6fe8232e7e265911bba0f632e178f86c3f0412c1265cf2962',
-     armv7l: 'efb6e1f5593f97b6fe8232e7e265911bba0f632e178f86c3f0412c1265cf2962',
-     x86_64: '2645a7f26f0d0f007c57f296ef18c853253b98da4756f7ada840a40147c8c8ed'
+    aarch64: '7327f9e37705d2b46a2a89300c154e16aa078f93bc3a6ce69932783ce6f56e50',
+     armv7l: '7327f9e37705d2b46a2a89300c154e16aa078f93bc3a6ce69932783ce6f56e50',
+     x86_64: '37c91128508324b3f865687061e46df5f650cc634489ec9d1a96b88f5917451a'
   })
 
   depends_on 'glibc_build237' => :build
-  depends_on 'glibc' # R
 
   conflicts_ok
   no_shrink
@@ -42,10 +41,18 @@ class Glibc_lib237 < Package
       FileUtils.install @filename_target, @destpath
     end
     # Symlinks to system libraries.
-    %w[libanl.so.1 libc_malloc_debug.so.0 libc.so.6 libdl.so.2 libm.so.6
-       libmvec.so.1 libnss_dns.so.2 libnss_files.so.2 libpthread.so.0
-       libresolv.so.2 librt.so.1 libthread_db.so.1 libutil.so.1].each do |lib|
-      FileUtils.ln_sf "/#{ARCH_LIB}/#{lib}", "#{CREW_DEST_LIB_PREFIX}/#{lib}"
+    @glibc_libs = %w[libanl.so.1 libc_malloc_debug.so.0 libc.so.6 libdl.so.2 libm.so.6
+                     libmvec.so.1 libnss_dns.so.2 libnss_files.so.2 libpthread.so.0
+                     libresolv.so.2 librt.so.1 libthread_db.so.1 libutil.so.1]
+    @glibc_libs -= ['libc.so.6'] if Kernel.system("patchelf --print-needed #{File.join(CREW_LIB_PREFIX, 'libc.so.6')} | grep -q libC.so.6")
+    @glibc_libs -= ['libm.so.6'] if Kernel.system("patchelf --print-needed #{File.join(CREW_LIB_PREFIX, 'libm.so.6')} | grep -q libC.so.6")
+
+    @glibc_libs.each do |lib|
+      if ARCH == 'x86_64'
+        puts "skipping symlink of system/#{ARCH_LIB}/#{lib} into #{CREW_DEST_LIB_PREFIX}/#{lib}...".orange
+      else
+        FileUtils.ln_sf "/#{ARCH_LIB}/#{lib}", "#{CREW_DEST_LIB_PREFIX}/#{lib}"
+      end
     end
   end
 end

--- a/packages/glibc_lib237.rb
+++ b/packages/glibc_lib237.rb
@@ -30,7 +30,7 @@ class Glibc_lib237 < Package
   def self.install
     puts 'Installing Glibc_build237 to pull files for build...'.lightblue
     @filelist_path = File.join(CREW_META_PATH, 'glibc_build237.filelist')
-    abort 'File list for Glibc_build237 does not exist!'.lightred unless File.file?(@filelist_path)
+    abort 'File list for Glibc_build237 does not exist! Please run: crew reinstall glibc_build237'.lightred unless File.file?(@filelist_path)
     @filelist = File.readlines(@filelist_path, chomp: true).sort
 
     @filelist.each do |filename|
@@ -40,24 +40,29 @@ class Glibc_lib237 < Package
       @filename_target = File.realpath(filename)
       FileUtils.install @filename_target, @destpath
     end
-    # Symlinks to system libraries.
+
+    if ARCH == 'x86_64'
+      puts 'Skipping creating symlinks from system Glibc libraries...'.orange
+      return
+    end
+
+    # Create symlinks to system glibc libraries.
     glibc_libs = %w[libanl.so.1 libc_malloc_debug.so.0 libc.so.6 libdl.so.2 libm.so.6
                     libmvec.so.1 libnss_dns.so.2 libnss_files.so.2 libpthread.so.0
                     libresolv.so.2 librt.so.1 libthread_db.so.1 libutil.so.1]
-    glibc_libs -= ['libc.so.6'] if Kernel.system("patchelf --print-needed #{File.join(CREW_LIB_PREFIX, 'libc.so.6')} | grep -q libC.so.6")
-    glibc_libs -= ['libm.so.6'] if Kernel.system("patchelf --print-needed #{File.join(CREW_LIB_PREFIX, 'libm.so.6')} | grep -q libC.so.6")
 
     glibc_libs.each do |lib|
-      if ARCH == 'x86_64'
-        puts "skipping symlink of system/#{ARCH_LIB}/#{lib} into #{CREW_DEST_LIB_PREFIX}/#{lib}...".orange
-      else
-        FileUtils.ln_sf "/#{ARCH_LIB}/#{lib}", "#{CREW_DEST_LIB_PREFIX}/#{lib}"
-      end
+      # Do not replace libraries that have been patched for our glibc libC.so.6.
+      next if Kernel.system("patchelf --print-needed #{File.join(CREW_LIB_PREFIX, lib)} | grep -q libC.so.6")
+
+      puts "Creating symlink of system Glibc's /#{ARCH_LIB}/#{lib} into #{CREW_DEST_LIB_PREFIX}/#{lib}...".orange
+      FileUtils.ln_sf "/#{ARCH_LIB}/#{lib}", "#{CREW_DEST_LIB_PREFIX}/#{lib}"
     end
   end
 
   def self.postinstall
     if (ARCH == 'x86_64') && (LIBC_VERSION.to_f >= 2.35)
+      # Overwrite our libc.so.6 with the system glibc libc.so.6, and then use patchelf to link it to ours.
       FileUtils.cp "/#{ARCH_LIB}/libc.so.6", File.join(CREW_LIB_PREFIX, 'libc.so.6.system') and FileUtils.mv File.join(CREW_LIB_PREFIX, 'libc.so.6.system'), File.join(CREW_LIB_PREFIX, 'libc.so.6') unless Kernel.system "patchelf --print-needed #{File.join(CREW_LIB_PREFIX, 'libc.so.6')} | grep -q libC.so.6"
       puts "patchelf is needed. Please run: 'crew install patchelf ; crew postinstall #{name}'".lightred unless File.file?(File.join(CREW_PREFIX, 'bin/patchelf'))
       # Link the system libc.so.6 to also require our renamed libC.so.6

--- a/packages/glibc_lib237.rb
+++ b/packages/glibc_lib237.rb
@@ -41,13 +41,13 @@ class Glibc_lib237 < Package
       FileUtils.install @filename_target, @destpath
     end
     # Symlinks to system libraries.
-    @glibc_libs = %w[libanl.so.1 libc_malloc_debug.so.0 libc.so.6 libdl.so.2 libm.so.6
+    glibc_libs = %w[libanl.so.1 libc_malloc_debug.so.0 libc.so.6 libdl.so.2 libm.so.6
                      libmvec.so.1 libnss_dns.so.2 libnss_files.so.2 libpthread.so.0
                      libresolv.so.2 librt.so.1 libthread_db.so.1 libutil.so.1]
-    @glibc_libs -= ['libc.so.6'] if Kernel.system("patchelf --print-needed #{File.join(CREW_LIB_PREFIX, 'libc.so.6')} | grep -q libC.so.6")
-    @glibc_libs -= ['libm.so.6'] if Kernel.system("patchelf --print-needed #{File.join(CREW_LIB_PREFIX, 'libm.so.6')} | grep -q libC.so.6")
+    glibc_libs -= ['libc.so.6'] if Kernel.system("patchelf --print-needed #{File.join(CREW_LIB_PREFIX, 'libc.so.6')} | grep -q libC.so.6")
+    glibc_libs -= ['libm.so.6'] if Kernel.system("patchelf --print-needed #{File.join(CREW_LIB_PREFIX, 'libm.so.6')} | grep -q libC.so.6")
 
-    @glibc_libs.each do |lib|
+    glibc_libs.each do |lib|
       if ARCH == 'x86_64'
         puts "skipping symlink of system/#{ARCH_LIB}/#{lib} into #{CREW_DEST_LIB_PREFIX}/#{lib}...".orange
       else

--- a/packages/glibc_lib237.rb
+++ b/packages/glibc_lib237.rb
@@ -42,8 +42,8 @@ class Glibc_lib237 < Package
     end
     # Symlinks to system libraries.
     glibc_libs = %w[libanl.so.1 libc_malloc_debug.so.0 libc.so.6 libdl.so.2 libm.so.6
-                     libmvec.so.1 libnss_dns.so.2 libnss_files.so.2 libpthread.so.0
-                     libresolv.so.2 librt.so.1 libthread_db.so.1 libutil.so.1]
+                    libmvec.so.1 libnss_dns.so.2 libnss_files.so.2 libpthread.so.0
+                    libresolv.so.2 librt.so.1 libthread_db.so.1 libutil.so.1]
     glibc_libs -= ['libc.so.6'] if Kernel.system("patchelf --print-needed #{File.join(CREW_LIB_PREFIX, 'libc.so.6')} | grep -q libC.so.6")
     glibc_libs -= ['libm.so.6'] if Kernel.system("patchelf --print-needed #{File.join(CREW_LIB_PREFIX, 'libm.so.6')} | grep -q libC.so.6")
 
@@ -57,14 +57,14 @@ class Glibc_lib237 < Package
   end
 
   def self.postinstall
-    if (ARCH == 'x86_64') && (LIBC_VERSION.to_f >= 2.35)
+    if (ARCH == 'x86_64') && (LIBC_VERSION >= 2.35)
       FileUtils.cp "/#{ARCH_LIB}/libc.so.6", File.join(CREW_LIB_PREFIX, 'libc.so.6.system') and FileUtils.mv File.join(CREW_LIB_PREFIX, 'libc.so.6.system'), File.join(CREW_LIB_PREFIX, 'libc.so.6') unless Kernel.system "patchelf --print-needed #{File.join(CREW_LIB_PREFIX, 'libc.so.6')} | grep -q libC.so.6"
       puts "patchelf is needed. Please run: 'crew install patchelf ; crew postinstall #{name}'".lightred unless File.file?(File.join(CREW_PREFIX, 'bin/patchelf'))
       # Link the system libc.so.6 to also require our renamed libC.so.6
       # which provides the float128 functions strtof128, strfromf128,
       # and __strtof128_nan.
       libc_patch_libraries = %w[libc.so.6 libm.so.6 libstdc++.so.6]
-      libc_patch_libraries.delete_if { |lib| !File.file?(File.join(CREW_LIB_PREFIX, lib)) }
+      libc_patch_libraries.keep_if { |lib| File.file?(File.join(CREW_LIB_PREFIX, lib)) }
       libc_patch_libraries.delete_if { |lib| Kernel.system "patchelf --print-needed #{File.join(CREW_LIB_PREFIX, lib)} | grep -q libC.so.6" }
 
       return if libc_patch_libraries.empty?

